### ISSUE TITLE
[Bug] performance in A10 for LLAMA2 -7B is slow

### ISSUE THREAD

**Issue by yongchaoding** [InternLM/lmdeploy#210]
### Checklist

- [X] 1. I have searched related issues but cannot get the expected help.
- [X] 2. The bug has not been fixed in the latest version.

### Describe the bug

I run LLAMA2-7B with Hugging Face in Single A10, the performance is about 34ms per token, and I test LMdeploy with batch size 1 equal with HF, and I found that the performance is about 27ms per token which is very slow. So If I miss some key point when using LMdeploy.

### Reproduction

```
# from transformers import AutoTokenizer, AutoModelForCausalLM
import torch
import time
import os
import os.path as osp
import random

from lmdeploy import turbomind as tm
from lmdeploy.model import MODELS
from lmdeploy.turbomind.tokenizer import Tokenizer

os.environ['TM_LOG_LEVEL'] = 'ERROR'

# model_path = "/workspace/inference/models/vicuna-7b-v1.3/workspace"
model_path = "/workspace/inference/models/Llama-2-7b-chat-hf/workspace"

# inputs torch.Size([1, 20])
# outputs torch.Size([1, 287])
# 287 - 20 = 267
prompt = ["Triton enables teams to deploy any AI model from multiple deep learning and machine learning frameworks"]

tokenizer_model_path = osp.join(model_path, 'triton_models', 'tokenizer')
tokenizer = Tokenizer(tokenizer_model_path)
tm_model = tm.TurboMind(model_path, eos_id=tokenizer.eos_token_id, tp=1)
generator = tm_model.create_instance()

nth_round = 1
step = 0
seed = random.getrandbits(64)
model_name = tm_model.model_name
model = MODELS.get(model_name)()

prompt = model.get_prompt(prompt, nth_round == 1)
prompt = tokenizer.encode(prompt)

repetition_penalty = 1.0
session_id = 1

# ## Inference
# warms up
for i in range(0, 1):
    outputs = []
    for output in generator.stream_infer(
                    session_id=session_id,
                    input_ids=[prompt],
                    stream_output=True,
                    request_output_len=1000,
                    sequence_start=(nth_round == 1),
                    sequence_end=False,
                    step=step,
                    stop=False,
                    top_k=50,
                    top_p=1.0,
                    temperature=1.0,
                    repetition_penalty=repetition_penalty,
                    ignore_eos=False,
                    random_seed=None):
        outputs.append(output)
    # outputs = llm.generate(prompt, sampling_params)
    # outputs = model.generate(inputs, max_new_tokens=1000, do_sample=False, top_k=50, top_p=0.95)

iter = 1
torch.cuda.synchronize()
start = time.time()
for i in range(0, iter):
    outputs = []
    result = []
    for output in generator.stream_infer(
                    session_id=session_id,
                    input_ids=[prompt],
                    stream_output=True,
                    request_output_len=1000,
                    sequence_start=(nth_round == 1),
                    sequence_end=False,
                    step=step,
                    stop=False,
                    top_k=50,
                    top_p=1.0,
                    temperature=1.0,
                    repetition_penalty=repetition_penalty,
                    ignore_eos=False,
                    random_seed=None):
        # outputs.append(output)
        res, tokens = output[0]
        result.append(res)
    # outputs = llm.generate(prompt, sampling_params)
    # outputs = model.generate(inputs, max_new_tokens=1000, do_sample=False, top_k=50, top_p=0.95)
torch.cuda.synchronize()
end = time.time()

# import pdb; pdb.set_trace()
print("Time: {} ms".format((end-start)*1000/iter))
# len(outputs[0].outputs[0].token_ids)
len_outputs = len(result)
print("Token: {} ms, Token Len: {}".format((end-start)*1000/iter/(len_outputs), len_outputs))

# Time: 3871.3719844818115 ms
# Token: 27.65265703201294 ms, Token Len: 140

response = tokenizer.decode(res)
print(response)
```

**Comment by lzhangzz** [comment]
We don't have A10 GPUs. However, here is some simple math

- According to official document, memory bandwidth of A10 is around 600GB/s.
- The model size of Llama-2 7B is 12.6GB. 

The inference speed of A10 on the 7B model is upper-bounded at 12.6 / 600 = 21ms/step. This is the time for a full sweep of model weights without doing anything else (which is required for every generation step). 

Taking other factors (attention kernels, slight under-utilization of DRAM bandwidth of GEMM kernels) into account, I guess it's reasonable to have 27.65ms/token on A10 with `batch_size = 1`.

To increase throughput, you may try to use a larger batch size so that the time for sweeping the weights is amortized by different samples in the batch.

**Comment by Dimensionzw** [comment]
> We don't have A10 GPUs. However, here is some simple math
> 
> * According to official document, memory bandwidth of A10 is around 600GB/s.
> * The model size of Llama-2 7B is 12.6GB.
> 
> The inference speed of A10 on the 7B model is upper-bounded at 12.6 / 600 = 21ms/step. This is the time for a full sweep of model weights without doing anything else (which is required for every generation step).
> 
> Taking other factors (attention kernels, slight under-utilization of DRAM bandwidth of GEMM kernels) into account, I guess it's reasonable to have 27.65ms/token on A10 with `batch_size = 1`.
> 
> To increase throughput, you may try to use a larger batch size so that the time for sweeping the weights is amortized by different samples in the batch.
> 
> You may also try #202, which can boost the speed by 2x on small batch sizes compared to 16-bit weights.

Hello, I would like to know whether the inference accuracy is reduced by using w4a16, and whether the output result is consistent with w16a16

**Comment by lzhangzz** [comment]
- Accuracy will most likely change on W4A16, it may decrease/increase on certain tasks. 
- There no consistency guarantee. In fact, with different GEMM (split-k vs non-split-k) kernels, consistency can not be maintained even using 16-bit weights.

Our implementation supports group-wise quantized INT4 weights (e.g. [AWQ](https://github.com/mit-han-lab/llm-awq) or [GPTQ](https://github.com/qwopqwop200/GPTQ-for-LLaMa) w/o reordering). Check out corresponding repos/papers for details.

AWQ's quantized weights is natively supported by our conversion script (GPTQ uses a transposed layout).
